src/Builders/PromptBuilder.php

```diff
@@ -1040,67 +1109,179 @@ private function getConfiguredModel(CapabilityEnum $capability): ModelInterface
-                            return $cap->value;
-                        }, $requirements->getRequiredCapabilities())),
-                        implode(', ', array_map(function ($opt) {
-                            return $opt->getName()->value . '=' . json_encode($opt->getValue());
-                        }, $requirements->getRequiredOptions()))
-                    )
+        if (empty($candidateMap)) {
+            $message = sprintf(
+                'No models found that support %s for this prompt.',
+                $capability->value
+            );
+
+            if ($this->providerIdOrClassName !== null) {
+                $message = sprintf(
+                    'No models found for provider "%s" that support %s for this prompt.',
+                    $this->providerIdOrClassName,
+                    $capability->value
                 );
             }
 
-            $firstProviderModels = $providerModelsMetadata[0];
-            $provider = $firstProviderModels->getProvider()->getId();
-            $modelMetadata = $firstProviderModels->getModels()[0];
-        } else {
-            $modelsMetadata = $this->registry->findProviderModelsMetadataForSupport(
-                $this->providerIdOrClassName,
-                $requirements
+            throw new InvalidArgumentException($message);
+        }
+
+        // Check if any preferred models match the candidates, in priority order.
+        $selectedModel = null;
+        if (!empty($this->modelPreferenceKeys)) {
+            // Find preferences that match available candidates, preserving preference order.
+            $matchingPreferences = array_intersect_key(
+                array_flip($this->modelPreferenceKeys),
+                $candidateMap
             );
 
-            if (empty($modelsMetadata)) {
-                throw new InvalidArgumentException(
-                    sprintf(
-                        'No models found for %s that support the required capabilities and options for this prompt. ' .
-                        'Required capabilities: %s. Required options: %s',
-                        $this->providerIdOrClassName,
-                        implode(', ', array_map(function ($cap) {
-                            return $cap->value;
-                        }, $requirements->getRequiredCapabilities())),
-                        implode(', ', array_map(function ($opt) {
-                            return $opt->getName()->value . '=' . json_encode($opt->getValue());
-                        }, $requirements->getRequiredOptions()))
-                    )
-                );
+            if (!empty($matchingPreferences)) {
+                // Get the first matching preference key
+                $firstMatchKey = key($matchingPreferences);
+                [$providerId, $modelId] = $candidateMap[$firstMatchKey];
+
+                $selectedModel = $this->registry->getProviderModel($providerId, $modelId, $this->modelConfig);
```

Comment:
I don't think we need the `$selectedModel` variable? We can just do an early return here.
```suggestion
                return $this->registry->getProviderModel($providerId, $modelId, $this->modelConfig);
```

Reply:
Good catch! Leftover from refactoring. Resolved in 2101ede